test-harness/src/ghciwatch.rs

```diff
@@ -190,35 +232,109 @@ impl GhciWatch {
+        chunk.push(event);
+        Ok(chunk.last().expect("We just inserted this event"))
+    }
+
+    /// Assert that a matching event was logged in one of the given `checkpoints`.
+    pub fn assert_logged_in_checkpoint(
+        &self,
+        checkpoints: impl CheckpointIndex,
+        matcher: impl IntoMatcher,
+    ) -> miette::Result<&Event> {
+        let mut matcher = matcher.into_matcher()?;
+        for event in self.events_in_checkpoints(checkpoints.clone()) {
+            if matcher.matches(event)? {
+                return Ok(event);
+            }
+        }
+
+        Err(miette!(
+            "No log message matching {matcher} found in checkpoint {:?}",
+            checkpoints.into_index()
+        ))
+    }
+
+    /// Assert that a matching event was logged since the last [`Checkpoint`].
+    pub fn assert_logged(&self, matcher: impl IntoMatcher) -> miette::Result<&Event> {
+        self.assert_logged_in_checkpoint(self.current_checkpoint(), matcher)
+    }
+
+    /// Match a log message in the given checkpoints or wait until a matching log event is
+    /// found.
+    ///
+    /// If `checkpoints` is `None`, do not check the `matcher` against any previously logged
+    /// events.
     ///
     /// Errors if waiting for the event takes longer than the given `timeout`.
-    pub async fn assert_logged_with_timeout(
+    pub async fn wait_for_log_with_timeout<M: IntoMatcher, C: CheckpointIndex>(
         &mut self,
-        negative_matcher: Option<Matcher>,
-        matcher: impl IntoMatcher,
+        matcher: M,
+        checkpoints: Option<C>,
         timeout_duration: Duration,
     ) -> miette::Result<Event> {
-        let matcher = matcher.into_matcher()?;
+        let mut matcher = matcher.into_matcher()?;
+
+        // First check if it was logged in `checkpoints`.
+        if let Some(checkpoints) = checkpoints {
+            if let Ok(event) = self.assert_logged_in_checkpoint(checkpoints, &mut matcher) {
+                return Ok(event.clone());
+            }
+        }
 
+        // Otherwise, wait for a log message.
         match tokio::time::timeout(timeout_duration, async {
             loop {
-                match self.tracing_reader.next_event().await {
-                    Err(err) => {
-                        return Err(err);
-                    }
-                    Ok(event) => {
-                        if matcher.matches(&event) {
-                            return Ok(event);
-                        } else if let Some(negative_matcher) = &negative_matcher {
-                            if negative_matcher.matches(&event) {
-                                return Err(miette!("Found a log event matching {negative_matcher}"));
-                            }
-                        }
-                    }
+                let event = self.read_event().await?;
+                if matcher.matches(event)? {
+                    return Ok(event.clone());
```

Comment:
(struggled to select the exact region I want in GitHub, sorry)

This block reads really strangely to me. I don't think the `Ok(Ok(event))` or `Ok(Err(err))` match arms are reachable. You'll either `return` early from the whole function, or you'll get `Err(_)` from the timeout. Otherwise you're just looping infinitely.

I would either:

1. Change the current `match the_block` to `if the_block.is_err()`
2. `break` from the loop with the matching event value instead of `return`ing from the function

I don't think your current code is incorrect, it's just weird lol.

Reply:
Oh, this is a horrible quirk of Rust's `async`/`await` compiling down into state machines: an `async` block like the one used here is like a separate function, so the `return` on line 337 only returns from the `async` block passed to `tokio::time::timeout`.

- The `Ok(Ok(event))` branch is reached from the `async` block return on line 337.
- The `Ok(Err(_))` branch is reached from the try (`?`) expressions on lines 335-336.
- The `Err(_)` branch is reached if the timeout expires.

Reply:
Ah, that makes sense! I knew about the async/await state machine stuff, but I didn't realize `return` was "scoped" to an `async` block. TIL.

Reply:
It's kind of evil because `async` is the only block scoped this way. There's been [a proposal for `try {}` blocks that behave similarly](https://doc.rust-lang.org/beta/unstable-book/language-features/try-blocks.html) since at least 2016 but there's some issues with type inference that have kept it from getting merged.